File: nodeapi/src/common/interceptors/logger.ts
import {
    CallHandler,
    ExecutionContext,
    Injectable,
    Logger,
    NestInterceptor,
    HttpException,
} from '@nestjs/common';
import { tap, catchError, throwError } from 'rxjs';
import { Request } from 'express';
import { User } from 'src/auth/entities/user.entity';

interface AuthenticatedRequest extends Request {
    user?: User;
}

@Injectable()
export class ActivityLogInterceptor implements NestInterceptor {
    private readonly logger = new Logger(ActivityLogInterceptor.name);

    intercept(context: ExecutionContext, next: CallHandler) {
        const request = context.switchToHttp().getRequest<AuthenticatedRequest>();
        const { method, path, user, headers, ip } = request;
        const userId = user?.id ?? 'Unauthenticated user';
        const xForwardedFor = (headers['x-forwarded-for'] ||
            'No Address') as string;
        const userAgent = headers['user-agent'] || 'No Agent';

        return next.handle().pipe(
            tap(() => {
                // Log 2xx responses
                this.logger.log(
                    `Response: ${method} ${path} - User ID: ${userId} - Status: 200 - Real IP : ${xForwardedFor} - User Agent: ${userAgent} - Remote IP : ${ip}`,
                );
            }),
            catchError((err: unknown) => {
                // Type guard for HttpException
                let statusCode = 500;
                let message = 'Internal Server Error';

                if (err instanceof HttpException) {
                    statusCode = err.getStatus();
                    message = err.message;
                } else if (err instanceof Error) {
                    message = err.message;
                }

                if (statusCode >= 500) {
                    this.logger.error(
                        `Server Error: ${method} ${path} - User ID: ${userId} - Status: ${statusCode} - Message: ${message} - Real IP : ${xForwardedFor} - User Agent: ${userAgent} - Remote IP : ${ip}`,
                    );
                } else if (statusCode >= 400) {
                    this.logger.warn(
                        `Client Error: ${method} ${path} - User ID: ${userId} - Status: ${statusCode} - Message: ${message} - Real IP : ${xForwardedFor} - User Agent: ${userAgent} - Remote IP : ${ip}`,
                    );
                }

                return throwError(() => err);
            }),
        );
    }
}

File: ai-backend/src/utils/logger.py
import logging
import time
from starlette.middleware.base import BaseHTTPMiddleware
from starlette.requests import Request
from starlette.responses import Response

logger = logging.getLogger("ActivityLogger")
logging.basicConfig(level=logging.INFO)


class ActivityLogMiddleware(BaseHTTPMiddleware):
    async def dispatch(self, request: Request, call_next) -> Response:
        method = request.method
        path = request.url.path
        xForwardedFor = request.headers.get("x-forwarded-for", "No Address")
        userAgent = request.headers.get("user-agent", "No Agent")
        ip = request.client.host if request.client else "unknown"
        userId = "Unauthenticated user"

        start = time.time()
        try:
            response: Response = await call_next(request)
            elapsed = round((time.time() - start) * 1000, 2)
            status = response.status_code
            if status >= 500:
                logger.error(
                    f"Server Error: {method} {path} - User ID: {userId} - Status: {status} - Real IP: {xForwardedFor} - User Agent: {userAgent} - Remote IP: {ip} - {elapsed}ms"
                )
            elif status >= 400:
                logger.warning(
                    f"Client Error: {method} {path} - User ID: {userId} - Status: {status} - Real IP: {xForwardedFor} - User Agent: {userAgent} - Remote IP: {ip} - {elapsed}ms"
                )
            else:
                logger.info(
                    f"Response: {method} {path} - User ID: {userId} - Status: {status} - Real IP: {xForwardedFor} - User Agent: {userAgent} - Remote IP: {ip} - {elapsed}ms"
                )
            return response
        except Exception as exc:
            logger.error(
                f"Exception: {method} {path} - User ID: {userId} - Real IP: {xForwardedFor} - User Agent: {userAgent} - Remote IP: {ip} - Error: {exc}"
            )
            raise
